### ISSUE TITLE
Require KeyCloak oauth  example

### ISSUE THREAD

**Issue by pzou19741** [dapr/dapr#2328]
## In what area(s)?

> /area runtime
> /area operator
> /area placement
> /area docs
> /area test-and-release

## Release Note

RELEASE NOTE:

**Comment by pzou19741** [comment]
any update?

**Comment by pzou19741** [comment]
I'm not able to make keycloak integration works.
I modified  https://github.com/dapr/quickstarts/tree/master/middleware  echoapp.

**Here is component yaml:**
```
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: oauth2
  namespace: default
spec:
  type: middleware.http.oauth2
  metadata:
  - name: clientId
    value: "test"
  - name: clientSecret
    value: "[keycloak client secret]"
  - name: scopes
    value: "openid"
  - name: authURL
    value: "https://[keycloak endpoint]/auth/realms/demo/protocol/openid-connect/auth"
  - name: tokenURL
    value: "https://[keycloak endpoint]/auth/realms/demo/protocol/openid-connect/token"
  - name: redirectURL
    value: "http://localhost:3500/v1.0/invoke/echoapp.default/"
  - name: authHeaderName
    value: "authorization" 
```
**configuration yaml**

```
apiVersion: dapr.io/v1alpha1
kind: Configuration
metadata:
  name: pipeline
  namespace: default
spec:
  httpPipeline:
    handlers:
    - type: middleware.http.oauth2
      name: oauth2  
```

deployment yaml 

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: echoapp
  namespace: default
  labels:
    app: echo
spec:
  replicas: 1
  selector:
    matchLabels:
      app: echo
  template:
    metadata:
      labels:
        app: echo
      annotations:
        dapr.io/enabled: "true"
        dapr.io/app-id: "echoapp"
        dapr.io/app-port: "3000"
        dapr.io/config: "pipeline"
    spec:
      containers:
      - name: echo
        image: archdevopsdevacr.azurecr.io/dapriosamples/middleware-echoapp:0.0.2
        ports:
        - containerPort: 3000
        imagePullPolicy: Always
      imagePullSecrets:
      - name: acrcred
```
I added console.log for debug   echoapp/app.js 

```

const express = require('express');
const bodyParser = require('body-parser');
const app = express();
app.use(bodyParser.json());

const daprPort = process.env.DAPR_HTTP_PORT || 3500;
const port = 3000;

app.get('/echo', (req, res) => {
    var text = req.query.text;
    console.log("Echoing: " + text);
    console.log("now let's print headers");
    for(var item in res.headers) {
	    console.log(item + ": " + res.headers[item]);
    }
    console.log("finished print   ");
    console.log("print headers: "+ res.headers);
    res.send("Access token: " + req.headers["authorization"] + " Text: " + text);


});

app.listen(port, () => console.log(`Node App listening on port ${port}!`));
```

**Comment by pzou19741** [comment]
I'm able to get Access Token via postman by   clientId, clientSecret, scope, authurl, tokenurl which defined on component yaml.
But when I 
 curl http://localhost:3500/v1.0/invoke/echoapp.default/method/echo?test
I got response:
Access token: undefined Text: undefined

**Comment by mchmarny** [comment]
You will need to provide the missing header parameters in curl like you do in Postman (client_id and client_password) using HTTP Basic Authorization.

**Comment by pzou19741** [comment]
@mchmarny  Thank you. Not sure if I understood correctly.  https://github.com/dapr/quickstarts/tree/master/middleware is oauth for google api.  I put all oauth required configuration on CRD component.    Also Step 5. only require put http://dummy.com/v1.0/invoke/echoapp/method/echo?text=hello on the browser..It did work for me for example.                    For Keycloak oauth, we are running Keycloak on same k8s cluster.    That's why I use dapr service invocation url.                             
          curl  http://localhost:3500/v1.0/invoke/echoapp.default/method/echo?test

**Comment by pzou19741** [comment]
@mchmarny  There is no missing parameters here.  I used same keycloak oauth parameters via post man and I'm able to get Access Token.    I thought Dapr will load all the configurations defined on component.yaml while invoking  "curl http://localhost:3500/v1.0/invoke/echoapp.default/method/echo?test"

**Comment by mchmarny** [comment]
So hold on, the same parameters you use in Postman do not work in curl? That has to be client side issue as it all looks the same on the server side.

**Comment by pzou19741** [comment]
@mchmarny I put the same list of parameters to component.yaml (configuration)  CRD.   I do not think I need to add same parameters to header  while invoke  curl http://localhost:3500/v1.0/invoke/echoapp.default/method/echo?test  

https://github.com/dapr/quickstarts/tree/master/middleware -> Diagram displayed.  Dapr OauthMiddleware will get from configuration and then talk to Oauth Server then push token header to app.

For KeyCloak OAuth Server, I do not see any logging there. But I do find logging while I'm using postman.   I tried to change  parameter's name from  oauth2 which is Component CRD and then redeploy the app .   for example I changed tokenURL to token_endpoint, etc. My app pod kept crash until I changed back.    

Here is KeyCloak openid_config
---------------------------
{"issuer":"https://auth.dev.archean.org/auth/realms/Archean","authorization_endpoint":"https://auth.dev.archean.org/auth/realms/Archean/protocol/openid-connect/auth","token_endpoint":"https://auth.dev.archean.org/auth/realms/Archean/protocol/openid-connect/token","token_introspection_endpoint":"https://auth.dev.archean.org/auth/realms/Archean/protocol/openid-connect/token/introspect","userinfo_endpoint":"https://auth.dev.archean.org/auth/realms/Archean/protocol/openid-connect/userinfo","end_session_endpoint":"https://auth.dev.archean.org/auth/realms/Archean/protocol/openid-connect/logout","jwks_uri":"https://auth.dev.archean.org/auth/realms/Archean/protocol/openid-connect/certs","check_session_iframe":"https://auth.dev.archean.org/auth/realms/Archean/protocol/openid-connect/login-status-iframe.html","grant_types_supported":["authorization_code","implicit","refresh_token","password","client_credentials"],"response_types_supported":["code","none","id_token","token","id_token token","code id_token","code token","code id_token token"],"subject_types_supported":["public","pairwise"],"id_token_signing_alg_values_supported":["PS384","ES384","RS384","HS256","HS512","ES256","RS256","HS384","ES512","PS256","PS512","RS512"],"id_token_encryption_alg_values_supported":["RSA-OAEP","RSA1_5"],"id_token_encryption_enc_values_supported":["A128GCM","A128CBC-HS256"],"userinfo_signing_alg_values_supported":["PS384","ES384","RS384","HS256","HS512","ES256","RS256","HS384","ES512","PS256","PS512","RS512","none"],"request_object_signing_alg_values_supported":["PS384","ES384","RS384","HS256","HS512","ES256","RS256","HS384","ES512","PS256","PS512","RS512","none"],"response_modes_supported":["query","fragment","form_post"],"registration_endpoint":"https://auth.dev.archean.org/auth/realms/Archean/clients-registrations/openid-connect","token_endpoint_auth_methods_supported":["private_key_jwt","client_secret_basic","client_secret_post","tls_client_auth","client_secret_jwt"],"token_endpoint_auth_signing_alg_values_supported":["PS384","ES384","RS384","HS256","HS512","ES256","RS256","HS384","ES512","PS256","PS512","RS512"],"claims_supported":["aud","sub","iss","auth_time","name","given_name","family_name","preferred_username","email","acr"],"claim_types_supported":["normal"],"claims_parameter_supported":false,"scopes_supported":["openid","microprofile-jwt","web-origins","roles","phone","address","email","profile","offline_access"],"request_parameter_supported":true,"request_uri_parameter_supported":true,"code_challenge_methods_supported":["plain","S256"],"tls_client_certificate_bound_access_tokens":true,"introspection_endpoint":"https://auth.dev.archean.org/auth/realms/Archean/protocol/openid-connect/token/introspect"}



Here is oauth related config:
----
  - name: clientId
    value: "test"
  - name: clientSecret
    value: "[keycloak client secret]"
  - name: scopes
    value: "openid"
  - name: authURL
    value: "https://[keycloak endpoint]/auth/realms/demo/protocol/openid-connect/auth"
  - name: tokenURL
    value: "https://[keycloak endpoint]/auth/realms/demo/protocol/openid-connect/token"
  - name: redirectURL
    value: "http://localhost:3500/v1.0/invoke/echoapp.default/"

**Comment by pzou19741** [comment]
Any update?

**Comment by pzou19741** [comment]
I'm able to make progress after deep into https://github.com/dapr/components-contrib/blob/master/middleware/http/ source code.

**Comment by yaron2** [comment]
> I'm able to make progress after deep into https://github.com/dapr/components-contrib/blob/master/middleware/http/ source code.

Can you share the progress you've made?

**Comment by pzou19741** [comment]
Sure.  Due to lack of Dapr logging.  I wrote on my golang and use "golang.org/x/oauth2" suggested by https://github.com/dapr/components-contrib/blob/master/middleware/http/ . I ran golang oauth code at my local and point to local Keycloak .   I  do not find Authorization  appear on request header  but everything worked as expected.    I felt Dapr oauth should be able to Keycloak.

This time I changed redirect url to httplocalhost:3000/echo but everything else remains unchanged
 as my original Dapr oauth example, Keycloack on k8s cluster.
Instead of using curl I using kubectl port-forward..
Here is detailed steps:
1)  kubectl port-forward pod/[echoapp podid] 3500:3500
2)  enter "http://localhost:3500/v1.0/invoke/echoapp.default/method/echo?text=TestKeycloak" on brower
3)  It redirect to Keycloak login page to use username/password.  I entered preconfiged keycloak user/name.
4)  It redirect to redirect url. (page can't be displayed. then   I did  kubectl port-forward pod/[echoapp podid] 3000:3000 
5) page displays "Access token: undefined Text: undefined"
